prusti-viper/src/encoder/procedure_encoder.rs

```diff
@@ -1130,160 +1138,262 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+                    mir_borrow_kind,
+                    place,
+                    location,
+                    encoded_lhs,
+                    ty
+                )?
+            }
+            &mir::Rvalue::Cast(mir::CastKind::Misc, ref operand, dst_ty) => {
+                self.encode_cast(
+                    operand,
+                    dst_ty,
+                    encoded_lhs,
+                    ty,
+                    location,
+                    span,
+                )?
+            }
+            &mir::Rvalue::Len(ref place) => {
+                self.encode_assign_sequence_len(
+                    encoded_lhs,
+                    place,
+                    ty,
+                    location,
+                )?
+            }
+            &mir::Rvalue::Repeat(ref operand, ref times) => {
+                self.encode_assign_array_repeat_initializer(
+                    encoded_lhs,
+                    operand,
+                    times,
+                    ty,
+                    location,
+                )?
+            }
+            ref rhs => {
+                unimplemented!("encoding of '{:?}'", rhs);
+            }
+        })
     }
 
     /// Encode the lhs and the rhs of the assignment that create the loan
     fn encode_loan_places(&mut self, loan_places: &LoanPlaces<'tcx>) -> EncodingResult<(vir::Expr, vir::Expr, bool, Vec<vir::Stmt>)> {
         debug!("encode_loan_places '{:?}'", loan_places);
+
         // will panic if attempting to encode unsupported type
-        let (expiring_base, mut stmts, expiring_ty, _) = self.encode_place(&loan_places.dest).unwrap();
+        let (expiring_base, mut stmts, expiring_ty, _) = self.encode_place(&loan_places.dest, ArrayAccessKind::Mutable(None, loan_places.location)).unwrap();
+        trace!("expiring_base: {:?}", (&expiring_base, &stmts, &expiring_ty));
+
+        // the original encoding of arrays is with a sort-of magic temporary variable, so
+        // `postprocess_place_encoding` will return `i32` here instead of Array$3$i32. so here
+        // we call the mir_encoder one that returns a `PlaceEncoding`, and do the
+        // postprocessing if necessary ourselves.
+        if let mir::Rvalue::Ref(_, mir::BorrowKind::Mut { .. }, ref rhs_place) = loan_places.source {
+            let (rhs_place_encoding, ..) = self.mir_encoder.encode_place(rhs_place).unwrap();
+            if let PlaceEncoding::ArrayAccess { .. } = rhs_place_encoding {
+                // encode expiry of the array borrow
+                let (expired_expr, regained_array, wand_rhs) = self.array_magic_wand_at[&loan_places.location].clone();
+
+                // expiring base is something like ref$i32, so we need .val_ref.val_int
+                let deref = self.encoder.encode_value_expr(expiring_base.clone(), expiring_ty);
+                let target_ty = if let ty::TyKind::Ref(_, target_ty, _) = expiring_ty.kind() { target_ty } else { unreachable!() };
```

Comment:
This doesn't panic but maybe nicer:
```suggestion
                let target_ty = expiring_ty.peel_refs();
```